x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -0,0 +1,274 @@
+	}
+	s.db = db
+	s.storage = New(db, ProtobufCodec{})
+	s.rw = s.storage.NewShardedReadWriter()
+	return nil
+}
+
+// Size returns the db size
+//
+// Caller should either be main Run loop or should be holding RLock already
+func (s *StorageManager) Size() (lsm, vlog int64) {
+	return s.db.Size()
+}
+
+func (s *StorageManager) runValueLogGC(discardRatio float64) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.db.RunValueLogGC(discardRatio)
+}
+
+// DropAndRecreate deletes the underlying badger DB at a file system level, and replaces it with a new badger DB.
+func (s *StorageManager) DropAndRecreate() {
+	s.mu.Lock()
+	s.rw.Close()
+	err := s.db.Close()
+	if err != nil {
+		s.logger.With(logp.Error(err)).Error("error closing badger db during drop and recreate")
+	}
+
+	s.subscriberPosMu.Lock()
+	backupPath := filepath.Join(filepath.Dir(s.storageDir), filepath.Base(s.storageDir)+".old")
+	err = os.Rename(s.storageDir, backupPath)
+	if err != nil {
+		s.logger.With(logp.Error(err)).Error("error renaming old badger db during drop and recreate")
+	}
+
+	// Since subscriber position file lives within badger DB directory,
+	// move back subscriber position file, as it is not a part of the DB
+	// FIXME: create s.storageDir first? Test this
+	err = os.Rename(filepath.Join(backupPath, subscriberPositionFile), filepath.Join(s.storageDir, subscriberPositionFile))
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		s.logger.With(logp.Error(err)).Error("error copying subscriber position file during drop and recreate")
+	}
+
+	err = s.Reset() //FIXME: this is likely fatal. Return error to crash the processor
+	if err != nil {
+		s.logger.With(logp.Error(err)).Error("error creating new badger db during drop and recreate")
+	}
+	s.subscriberPosMu.Unlock()
+	s.mu.Unlock()
+
+	err = os.RemoveAll(backupPath)
+	if err != nil {
+		s.logger.With(logp.Error(err)).Error("error removing old badger db during drop and recreate")
+	}
+}
+
+func (s *StorageManager) ReadSubscriberPosition() ([]byte, error) {
+	s.subscriberPosMu.Lock()
+	defer s.subscriberPosMu.Unlock()
+	return os.ReadFile(filepath.Join(s.storageDir, subscriberPositionFile))
+}
+
+func (s *StorageManager) WriteSubscriberPosition(data []byte) error {
+	s.subscriberPosMu.Lock()
+	defer s.subscriberPosMu.Unlock()
+	return os.WriteFile(filepath.Join(s.storageDir, subscriberPositionFile), data, 0644)
+}
+
+func (s *StorageManager) NewReadWriter() *ManagedReadWriter {
+	return &ManagedReadWriter{
+		sm: s,
+	}
+}
+
+// ManagedReadWriter is a read writer that is transparent to underlying badger DB swaps done by StorageManager
+type ManagedReadWriter struct {
+	sm *StorageManager
+}
+
+func (s *ManagedReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
+	s.sm.mu.RLock()
+	defer s.sm.mu.RUnlock()
+	return s.sm.rw.ReadTraceEvents(traceID, out)
+}
+
+func (s *ManagedReadWriter) WriteTraceEvent(traceID, id string, event *modelpb.APMEvent, opts WriterOpts) error {
+	s.sm.mu.RLock()
```

Comment:
[to reviewer] not sure if it should RLock and wait for the swap, or just fail right away with TryRLock to avoid blocking ingestion.

Reply:
I think this is okay for now. If we observe contention then we can revisit it.